src/test/java/org/jenkinsci/plugins/workflow/cps/DeserializationTest.java

```diff
@@ -0,0 +1,136 @@
+package org.jenkinsci.plugins.workflow.cps;
+
+import groovy.lang.GroovyShell;
+import hudson.model.Result;
+import org.jenkinsci.plugins.scriptsecurity.scripts.ScriptApproval;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+import javax.annotation.CheckForNull;
+
+public class DeserializationTest {
+    @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    private static final String EXISTING_VAR_NAME = "INJECTED_VAR";
+    private static final String EXISTING_VAR_VALUE = "PRE_EXISTING";
+    private static final String A_GROOVY = "a.groovy";
+    private static final String A_CONTENTS = "def call(arg) {echo \"a ran on ${arg}\"}; this";
+    private static final String B_GROOVY = "b.groovy";
+    private static final String B_CONTENTS = "def m(arg) {" +
+                                             "echo \"${this} binding=${binding.variables}\";" +
+                                             "a(\"${arg} from b\")}; this";
+    private static String LOAD_PREFIX = "a = 0;" +
+                                        "node {a = load 'a.groovy'};" +
+                                        "echo \"Pre-semaphore value is ${" + EXISTING_VAR_NAME + "}\";";
+    private static String LOAD_SEMAPHORE = "semaphore 'wait';";
+    private static String LOAD_SUFFIX = "echo \"Post-semaphore value is ${" + EXISTING_VAR_NAME + "}\";" +
```

Comment:
nit: I find it really hard to read tests where the Pipeline scripts is stored in fragments in fields like this, since you have to jump around the file to see what is happening. I prefer inlining the scripts into the tests so you can see everything at once even if it means there is a bit of duplication.

Also, one thing that can be useful to avoid some escaping is that Groovy supports `/` as a `GString` delimiter with the same behavior as `"`, so you could write this instead:

```suggestion
    private static String LOAD_SUFFIX = "echo(/Post-semaphore value is ${" + EXISTING_VAR_NAME + "}/)" +
```

Reply:
Nice tip! I thought that was a typo.
Fair point, I'll go ahead and inline the scripts. I mostly did it because I had the opposite problem where I was trying to see what the difference was between the control and the test script.